[Data Cleaning] Address any concurrency issues (or potential issues) and performance imrovements

biyeun opened the pull request:
## Technical Summary
In an HTMX world, we need to always be defensive against and robustly handle concurrency issues with the db because they absolutely will happen.

This PR addresses any potential concurrency issues in data cleaning models and implements a retry in case any action results in an `IntegrityError`.

should address this 🐞 
https://dimagi.atlassian.net/browse/SAAS-17358

Additionally, I've made a couple small performance improvements.


## Feature Flag
`DATA_CLEANING_CASES`

## Safety Assurance

### Safety story
very safe change. tests still pass and all of changed logic is tested. only affects feature flagged workflows

### Automated test coverage
yes

### QA Plan
not yet


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

biyeun committed:
fix concurrency issues with select_record

biyeun committed:
fix potential concurrency issues on select_multiple_records

biyeun committed:
defensive strategy: re-update any records that might still not be marked if there were any conflicts

biyeun committed:
reduce concurrency issues for restart_case_session

biyeun committed:
reduce db interactions with _update_order

biyeun committed:
improve concurrency issue handling for add_filter

biyeun committed:
improve concurrency issue handling for deselect record

biyeun committed:
explicitly return None (cleanup)

biyeun committed:
improve concurrency issue handling for deselect multiple records

biyeun committed:
improve concurrency issue handling for remove column

biyeun committed:
improve concurrency issue handling for remove filter

nospame commented on corehq/apps/data_cleaning/utils/decorators.py:
```diff
@@ -0,0 +1,20 @@
+                try:
+                    return func(*args, **kwargs)
+                except IntegrityError:
+                    if attempt < max_retries - 1:
```
Using the default of 3, this would mean there are only 3 tries total before we hit the `raise`, right (attempt 0, 1, and 2)? Should this just be `if attempt < max_retries`?

biyeun replied:
If you changed the condition to if attempt < max_retries:, then for each iteration (0, 1, and 2) the condition would always be true—because 0, 1, and 2 are all less than 3. That would prevent the error from ever being raised, because the loop would never hit the else clause.

It **does** retry the function 3 times, just on the 3rd time we want it to raise the error if it encounters it.

nospame replied:
Right, that's on me forgetting that `range` is exclusive of the final number, not inclusive -- which makes sense. But doesn't that still mean this will only be tried 3 times total, rather than _retried_ 3 times after the initial attempt?

biyeun replied:
it will be tried 3 times total yes... I see what you are saying

nospame commented on corehq/apps/data_cleaning/models.py:
```diff
@@ -859,6 +859,13 @@ def select_multiple_records(cls, session, doc_ids):
 
+        # re-update any records that might still not be marked if there
+        # were any conflicts above...
+        session.records.filter(
```
Instead of trying again here after using `ignore_conflicts=True`, could we leave `ignore_conflicts` unset and rely on `retry_on_integrity_error` to handle the retries instead? Or are those things handling different potential issues?

biyeun replied:
They do address different concerns. In this case, setting `ignore_conflicts=True` on `bulk_create` is a targeted way to handle the situation where another process has already inserted a record—allowing the current `bulk_create` to quietly skip those conflicting rows without raising an exception. This avoids aborting the whole transaction and having the retry decorator restart the entire function.

If we left `ignore_conflicts` unset, then an `IntegrityError` would be raised when a conflict occurs, which would trigger the `retry_on_integrity_error` decorator. That would cause the entire function to be retried from the start. While that may eventually get to a state where the records are properly marked as selected, it can be less efficient because we are re-executing all the queries inside the function rather than just handling the expected, benign conflict locally.

nospame replied:
That makes sense. I don't love that it's just lines 876-879 repeated again, but I don't have a better suggestion, and I don't think it's blocking.

biyeun committed:
update tests to match input format from forms

biyeun committed:
fix UI errors -- avoid UUID vs string comparison
...forms will past a list of strings, while remove_column / remove_filter passes a list of UUIDs. Avoid UUID comparison by casting the id in the instance_map to a string

biyeun committed:
update retry_on_integrity_error with note
...explaining that the zero-th attempt is not part of the retry count.

millerdev commented on corehq/apps/data_cleaning/utils/decorators.py:
```diff
@@ -0,0 +1,20 @@
+from django.db import IntegrityError
+
+
+def retry_on_integrity_error(max_retries=3, delay=0.1):
```
We already have a [`retry_on` decorator](https://github.com/dimagi/commcare-hq/blob/a9edcd7e3e75d3213ed08e2151fd6c42c25380f6/corehq/ex-submodules/dimagi/utils/retry.py) that does something similar to this, but with exponential backoff. You could use it to create a `retry_on_integrity_error` similar to this example:

https://github.com/dimagi/commcare-hq/blob/a9edcd7e3e75d3213ed08e2151fd6c42c25380f6/corehq/ex-submodules/dimagi/utils/couch/database.py#L173-L177

nospame reviewed (approved):
Curious about your thoughts on @millerdev 's comment about building `retry_on_integrity_error` on the existing `retry_on` decorator, but overall looks good to me.

biyeun commented:
@nospame that looks ideal, i will build on that in a followup PR. thanks @millerdev
